### DISCUSSION TITLE
Cannot import mysqldump due to order of tables and foreign keys

### DISCUSSION THREAD

**Discussion by vincentrolfs** [planetscale/discussion#570] [Bug Reports]
My dump.sql come from mysqldump, and begins similar to this:

```sql
/*!40101 SET @OLD_CHARACTER_SET_CLIENT=@@CHARACTER_SET_CLIENT */;
/*!40101 SET @OLD_CHARACTER_SET_RESULTS=@@CHARACTER_SET_RESULTS */;
/*!40101 SET @OLD_COLLATION_CONNECTION=@@COLLATION_CONNECTION */;
/*!50503 SET NAMES utf8mb4 */;
/*!40103 SET @OLD_TIME_ZONE=@@TIME_ZONE */;
/*!40103 SET TIME_ZONE='+00:00' */;
/*!40014 SET @OLD_UNIQUE_CHECKS=@@UNIQUE_CHECKS, UNIQUE_CHECKS=0 */;
/*!40014 SET @OLD_FOREIGN_KEY_CHECKS=@@FOREIGN_KEY_CHECKS, FOREIGN_KEY_CHECKS=0 */;
/*!40101 SET @OLD_SQL_MODE=@@SQL_MODE, SQL_MODE='NO_AUTO_VALUE_ON_ZERO' */;
/*!40111 SET @OLD_SQL_NOTES=@@SQL_NOTES, SQL_NOTES=0 */;

--
-- Table structure for table `admintrainings`
--

DROP TABLE IF EXISTS `admintrainings`;
/*!40101 SET @saved_cs_client     = @@character_set_client */;
/*!50503 SET character_set_client = utf8mb4 */;
CREATE TABLE `admintrainings` (
  `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
  `training_id` int(10) unsigned NOT NULL,
  CONSTRAINT `admintrainings_training_id_foreign` FOREIGN KEY (`training_id`) REFERENCES `trainings` (`id`) ON DELETE CASCADE ON UPDATE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=124 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;
/*!40101 SET character_set_client = @saved_cs_client */;
```

The admintrainings table references trainings with a foreign key. But trainings does not exist yet at that point. As far as I can tell, that is pretty standard for mysqldump, and everything should still work due to `FOREIGN_KEY_CHECKS=0`.

Unfortunately, upon importing the dump (for example using SOURCE), I get an error:
```
ERROR 1824 (HY000) at line 25: target: mydatabase.-.primary: vttablet: rpc error: code = Unknown desc = Failed to open the referenced table 'trainings'
```

What can be done? Thanks!

**Comment by orware** [comment]
Hello @vincentrolfs,

Our team has seen your post above and is investigating this on our end for possible long-term solutions to this particular issue.

In my quick tests earlier while I did observe the errors you shared, I also noticed that if I ran the import twice via the `source` command that the first attempt would bring in the parent tables, while the second attempt would bring in the child tables. I still received some error messages in the output but they were due to that codependency amongst the tables due to the constraints.

Right now, I don't think that the `FOREIGN_KEY_CHECKS=0` setting is being applied as expected when importing via this path so that may be the explanation for the current behavior. Additionally, since the `mysqldump` output doesn't apply an ordering to the tables when it generates the dump file so that the parent tables are listed first, it instead seems to rely on the `FOREIGN_KEY_CHECKS=0` setting adjustment to prevent these types of errors when importing data again from a dump file. 

I would be curious if the two-step approach mentioned above about running `source` twice could function as a workaround for your situation.

Another option would be to place parent tables, such as the `trainings` one, at the beginning of your file before your child tables, depending on how many of these foreign key references are within your dump file. If there aren't that many or the dataset isn't that large, this may also provide a relatively quick solution.

But rest assured that what you have reported here is a valid issue and we'll be taking a closer look to see what we may be able to do to help with these.

If you'd like to work with us more closely on a resolution, please reach out via https://support.planetscale.com/ and it will make it easier to coordinate and run other tests together with you.

**Comment by dbussink** [comment]
A fix for this is in the works in https://github.com/vitessio/vitess/pull/14696.

**Reply by dbussink** [reply]
@vincentrolfs Can you please contact https://support.planetscale.com/ with the details about which database etc. you're testing with?
